open_feature/open_feature_client.py

```diff
@@ -149,34 +149,65 @@ def evaluate_flag_details(
         flag_type: FlagType,
         key: str,
         default_value: typing.Any,
-        evaluation_context: typing.Any = None,
+        evaluation_context: EvaluationContext = None,
         flag_evaluation_options: typing.Any = None,
     ) -> FlagEvaluationDetails:
+        if evaluation_context is None:
+            evaluation_context = EvaluationContext()
         try:
-            args = (
+            # The merge below will eventually pull in Evaluation contexts from Hooks
+            invocation_context = EvaluationContext()
+            invocation_context.merge(ctx1=EvaluationContext(), ctx2=evaluation_context)
+            return self.create_provider_evaluation(
+                flag_type,
                 key,
                 default_value,
-                evaluation_context,
+                invocation_context,
                 flag_evaluation_options,
             )
-            if flag_type is FlagType.BOOLEAN:
-                return self.provider.get_boolean_details(*args)
-
-            elif flag_type is FlagType.NUMBER:
-                return self.provider.get_number_details(*args)
-
-            elif flag_type is FlagType.OBJECT:
-                return self.provider.get_object_details(*args)
-            # Fallback case is a string object
-            elif flag_type is FlagType.STRING:
-                return self.provider.get_string_details(*args)
-            else:
-                raise GeneralError(error_message="Unknown flag type")
-
-        except OpenFeatureError:
+        except OpenFeatureError as ofe:
             return FlagEvaluationDetails(
                 key=key,
                 value=default_value,
                 reason=Reason.ERROR,
-                error_code=ErrorCode.GENERAL,
+                error_code=ofe.error_message,
             )
+
+    def create_provider_evaluation(
+        self,
+        flag_type: FlagType,
+        key: str,
+        default_value: typing.Any,
+        evaluation_context: EvaluationContext = None,
+        flag_evaluation_options: typing.Any = None,
+    ) -> FlagEvaluationDetails:
+        """
+        Encapsulated method to create a FlagEvaluationDetail from a specific provider.
+
+        :param flag_type: the type of the flag being returned
+        :param key: the string key of the selected flag
+        :param default_value: backup value returned if no result found by the provider
+        :param evaluation_context: Information for the purposes of flag evaluation
+        :param flag_evaluation_options: Additional flag evaluation information
+        :return: a FlagEvaluationDetails object with the fully evaluated flag from a
+        provider
+        """
+        args = (
+            key,
+            default_value,
+            evaluation_context,
+            flag_evaluation_options,
+        )
+        if flag_type is FlagType.BOOLEAN:
```

Comment:
Perhaps a slightly more pythonic (and succinct) way of writing this would be: 

```python
get_details_callable = {
    FlagType.BOOLEAN: self.provider.get_boolean_details,
    FlagType.NUMBER: self.provider.get_number_details,
    ...
}.get(flag_type)
if not get_details_callable:
    raise GeneralError(...)
return get_details_callable(*args)
```